File: app/game_engine/boards/square_board.rb
class SquareBoard

  attr_reader :rows, :columns

  def initialize(rows, columns)
    @rows = rows
    @columns = columns
  end

  def coordinate_valid?(coordinate)
    0 <= coordinate['x'] && coordinate['x'] < columns && 0 <= coordinate['y'] && coordinate['y'] < rows
  end

  def center_coordinate
    { 'x' => columns / 2, 'y' => rows / 2 }
  end

  def territory(coordinate)
    if rows.odd? && coordinate['y'] == rows / 2
      'neutral'
    elsif coordinate['y'] < rows / 2
      'alabaster'
    else
      'onyx'
    end
  end

  def directional_functions(type)
    if type == 'orthogonal'
      [
        lambda{ |coordinate| coordinate['x'] += 1},
        lambda{ |coordinate| coordinate['x'] -= 1},
        lambda{ |coordinate| coordinate['y'] += 1},
        lambda{ |coordinate| coordinate['y'] -= 1}
      ]
    elsif type == 'diagonal'
      [
        lambda{ |coordinate| coordinate['x'] += 1; coordinate['y'] += 1},
        lambda{ |coordinate| coordinate['x'] += 1; coordinate['y'] -= 1},
        lambda{ |coordinate| coordinate['x'] -= 1; coordinate['y'] += 1},
        lambda{ |coordinate| coordinate['x'] -= 1; coordinate['y'] -= 1}
      ]
    else
      raise "#{self.class}#directional_functions does not support type: #{type}"
    end
  end

  def distance(coordinate1, coordinate2)
    x_diff = coordinate1['x'] - coordinate2['x']
    y_diff = coordinate1['y'] - coordinate2['y']

    Math.sqrt(x_diff ** 2 + y_diff ** 2)
  end
end

File: api/src/services/game/board/square_board.ts
import { ICoordinate } from "../../../shared/dtos/game";
import {
  ICoordinateUpdater,
  BoardDirection,
  SetupTerritoryOwner,
  IBoard,
} from ".";

export class SquareBoard implements IBoard {
  private readonly columns: number;
  private readonly rows: number;

  constructor(columns: number, rows: number) {
    this.columns = columns;
    this.rows = rows;
  }

  getAllCoordinates(): ICoordinate[] {
    const result: ICoordinate[] = [];
    for (let x = 0; x <= this.columns; x++) {
      for (let y = 0; y <= this.rows; y++) {
        result.push({ x, y });
      }
    }
    return result;
  }

  getCenter(): ICoordinate {
    return { x: Math.floor(this.columns / 2), y: Math.floor(this.rows / 2) };
  }

  getCoordinateDistance(
    coordinate1: ICoordinate,
    coordinate2: ICoordinate
  ): number {
    const xDiff = coordinate1.x - coordinate2.x;
    const yDiff = coordinate1.y - coordinate2.y;
    return Math.sqrt(Math.pow(xDiff, 2) + Math.pow(yDiff, 2));
  }

  getDirectionalFunctions(type: BoardDirection): ICoordinateUpdater[] {
    if (type === BoardDirection.orthogonal) {
      return [
        (coordinate) => ({ x: coordinate.x + 1, y: coordinate.y }),
        (coordinate) => ({ x: coordinate.x - 1, y: coordinate.y }),
        (coordinate) => ({ x: coordinate.x, y: coordinate.y + 1 }),
        (coordinate) => ({ x: coordinate.x, y: coordinate.y - 1 }),
      ];
    }
    if (type === BoardDirection.diagonal) {
      return [
        (coordinate) => ({ x: coordinate.x + 1, y: coordinate.y + 1 }),
        (coordinate) => ({ x: coordinate.x - 1, y: coordinate.y - 1 }),
        (coordinate) => ({ x: coordinate.x + 1, y: coordinate.y - 1 }),
        (coordinate) => ({ x: coordinate.x - 1, y: coordinate.y + 1 }),
      ];
    }
    throw Error("Unsupported board direction");
  }

  getSetupTerritoryOwner(coordinate: ICoordinate): SetupTerritoryOwner {
    if (
      Math.floor(this.rows / 2) !== this.rows / 2 &&
      coordinate.y === Math.floor(this.rows / 2)
    ) {
      return SetupTerritoryOwner.neutral;
    }
    if (coordinate.y < this.rows / 2) {
      return SetupTerritoryOwner.alabaster;
    }
    return SetupTerritoryOwner.onyx;
  }

  isCoordinateValid(coordinate: ICoordinate): boolean {
    return (
      coordinate.x >= 0 &&
      coordinate.x < this.columns &&
      coordinate.y >= 0 &&
      coordinate.y < this.rows
    );
  }
}
